Search improvements

abaevbog opened the pull request:
1. In advanced search, if 'collection' condition is present, return and display in itemTree item's children that match the query even if
parents do not match. [Brief demo video](https://www.dropbox.com/scl/fi/9ovwp64q5qskyiormwuxy/collections_match.mov?rlkey=6nv0p1d7usng9mxrgwsv12ers&dl=0). Fixes: #1838 

2. Fix to the bug that caused the conditions to be ignored when a search condition is removed and re-added. Fixes: #3434 

3. When search is opened and itemTree notifier receiver `item` `modify` message, do not do full reload. Instead, run a search with existing scope on given itemIDs and add/remove rows as needed depending on the output of the search. It makes longer operations, such as adding tags, faster for large libraries.

abaevbog committed:
keep search conditionIDs in arithmetic sequence

When conditions are removed, shift conditionIDs so that
conditionIDs always go in increments of 1 (0, 1, 2 ...).
It prevents conditionIDs from conflicting with each other
when conditions are rearranged.

Fixes: #3434

abaevbog committed:
match child items during collection search

When a collection condition is present in advanced search,
child items are matched in addition to top-level items.

Fixes: #1838

abaevbog commented:
@dstillman, this PR has a fix to the advanced search issues that were very recently discussed in the [forums](https://forums.zotero.org/discussion/comment/448899#Comment_448899). 

[4ea9be9](https://github.com/zotero/zotero/pull/3435/commits/4ea9be96458c08b477e9ffbb777862684f4d2bfa) should address the issue with some conditions not being properly executed as well as the minus sign not removing the condition. There is a simple test that will fail if ran on master without these changes.

Two other commits are improvements that may be less important at this point - would it be better to pull [4ea9be9](https://github.com/zotero/zotero/pull/3435/commits/4ea9be96458c08b477e9ffbb777862684f4d2bfa) into a separate PR to look at it by itself?

dstillman commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -569,6 +566,77 @@ var ItemTree = class ItemTree extends LibraryTree {
+					let rowIndex = this.getRowIndexByID(id);
+					let item = Zotero.Items.get(id);
+					// If a row exists but it's item is not in search results,
+					// it should be removed unless it's children match the search
```
`it's` → `its` x 2

dstillman reviewed (commented):
The 'modify' optimization seems to be moving a matching child note when I edit another child note in a saved search ("Note" "contains" "foo" and "Include parent and child items of matching items").

Before:

![Screenshot 2023-11-21 at 7 23 11 AM](https://github.com/zotero/zotero/assets/783082/62b076a4-092a-4153-b1c2-428b7307adc7)

After modifying the `Annotations (24/04/2023 à 21:29:02)` note:

![Screenshot 2023-11-21 at 7 23 48 AM](https://github.com/zotero/zotero/assets/783082/382c7864-c149-457d-82b4-43791910e0ef)

The `Annotations (1/19/2023, 7:47:19 PM)` child note, which hasn't been modified, is moved to the top level below the sibling items of the note I modified.

dstillman commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -569,6 +568,77 @@ var ItemTree = class ItemTree extends LibraryTree {
+					else if (searchResultsSet.has(id) && rowIndex === false) {
+						let shouldOpen = false;
+						if (item.parentItemID) {
+							// Check if it's parent's row exists
```
`its`

abaevbog commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -369,6 +369,8 @@ var ItemTree = class ItemTree extends LibraryTree {
 			// TODO: Only refresh on condition change (not currently available in extraData)
 			await this.refresh();
+			// Redraw the tree to remove any old rows that do not match the filter
+			this.tree.invalidate();
```
I am still tracking down exactly the best way to tackle this issue I ran into and I don't think I have a 100% good solution yet but here are my findings.

1. This is a separate issue from the ordering discussed earlier.
2. The issue can be seen by taking any saved search with 2 conditions in "match any" mode and removing one of the conditions. Without clicking on anything else, the rows that match the removed condition will still stick around until you click something, and only then they go away.
3. Adding a new condition to saved search will (sometimes?) add rows but display them as context rows until you click around, in which case they stop being greyed outr. 
4. Adding a condition to the saved search that matches a child of an already matching collapsed container row does not expand it until you click around. Same happens the other way around when a child stops matching and the container needs to be collapsed.
5. Note that this only happens when there is not selection in itemTree - if any row in itemTree is selected, things works as expected.
6. Also sometimes this does not seem to happen, so there may be some kind of race condition involved.


As I see, part of the issue can be that the tree invalidation in `sort()` that `refresh` calls is never reached. In our case of removing a saved search condition, `sort` [returns](https://github.com/zotero/zotero/blob/436b19d64cb942644ea80705b5247dc52e6ebeb7/chrome/content/zotero/itemTree.jsx#L1243) before getting to the tree invalidation, so I think the tree is just never updated in that case.

I just pushed a commit that seems to partly address these things. Checking if `sort()` ran and invalidated the tree works to remove the non-matching rows on condition change but does not (always) help with displaying newly matching children

abaevbog replied:
From my experimentation so far, just keeping `this.tree.invalidate();` after  `await this.refresh();`  on `search` `modify` notification seems to solve all of the issues above, unlike invalidating the tree from inside of `this.refresh` as in my current last commit. I assume it has something to do with when the tree invalidation runs

abaevbog replied:
To address number 4 from above,  we could do something like
```
if (type == 'search' && action == 'modify') {
	// TODO: Only refresh on condition change (not currently available in extraData)
	await this.refresh();
	// Redraw the tree to remove any old rows that do not match the filter
        if (!this.selection.focused) {
            this.tree.invalidate();
        }
}
```
That way, it would only run when there is no selection, which is when these issues (most importantly number 4 that's no handled by adding manual tree invalidation in refresh) happen. But it's not a fantastic solution since it doesn't address the actual reason of why it happens.

abaevbog replied:
From logs, i see that there are calls to invalidate the tree but `render` in itemTree is not called the last time with all the rows added to it after refresh. For some reason, invalidating the tree a bit after `refresh` as I did earlier (or even manually in Run Javascript) does result in `itemTree.render` being called.  Where can I find the mechanism connecting the tree invalidation and call to `render` in itemTree? I thought one always follows the other but that's clearly not the case.

Also, I have a [brief video](https://www.dropbox.com/scl/fi/j4chxg19bkai527smv9v1/saved_search_rendering.mov?rlkey=eqbznsfq3emfml268nwk3ea0w&dl=0) of the issue, just to make sure it's clear what I'm looking at

abaevbog replied:
This issue does seem to boil down to the tree not being invalidated in a few very specific situations when no item is selected in itemTree of saved search: a) when items should only be removed after, say, a condition update that only matches as subset of current items, so nothing is done in `sort()` where tree invalidation would happen otherwise b) when a parent needs to be expanded to reveal a matching child, `_restoreSelection`(which would invalidate rows by toggling them) returns right away if there is no selected item. 

I re-pushed my earlier changes to only invalidate the tree in `refresh` in those specific situations and it seems to address this for me

abaevbog commented:
> The 'modify' optimization seems to be moving a matching child note when I edit another child note in a saved search ("Note" "contains" "foo" and "Include parent and child items of matching items").

This is ... very odd. I have not been able to reproduce it yet - does it happen regularly or every once in a while? Also, does this happen only in saved search and not in advanced search? And the moved annotation belonged to a completely different item, right?

I had a few things off that could have affected the overall saved search setup - I was using `collectionTree.getSearchObject()` instead of `collectionTree.ref`, which was not giving me the correct search object (not sure why), so some rows would not get removed properly from the saved search. Also, I had a few synchronous `toggleOpenState` calls that were not skipping map refresh, so that could cause a kind of race condition during re-rendering too. 

These could have contributed to this behavior.

abaevbog commented:
Another report that boils down to the collection search condition not matching child items:

https://forums.zotero.org/discussion/122782/bug-or-unsupported-feature-batch-rename-pdf-attachments-in-the-same-collection

dstillman commented:
Do we need to update this to handle annotations?

(https://forums.zotero.org/discussion/comment/491267/#Comment_491267 notes that "Include parent and child items of matching items" no longer works when you're searching for annotations. I guess there's a separate question of whether that option needs to be expanded to include all levels.)

abaevbog committed:
itemTree performance optimization in search mode

When an item is modified in search mode, do not rerun
the entire query and refresh the entire search list.
Instead, run the query on the modified items and manually
add/remove/expand/collapse item rows depending on if
updated items and their parents/children should still
be present in search results.

abaevbog committed:
search updates for annotation rows

- annotations will match collection conditions
- allow 'isNot' conditions to match annotation items
- includeParentsAndChildren will account for annotations
- cleanup of comments and tests remaining from when
annotation conditions would match attachments

abaevbog committed:
use itemTree optimization w quickSearch and tags

And minor edge cases fixes

abaevbog committed:
handle standalone <-> child item updates

In itemTree optimization

abaevbog commented:
I believe now this has the behavior we are after. This PR ended up having a whole number of search-related fixes, so it's a bit messy, sorry about that. To sum up:
- the issue of conditions not properly applying after they are removed and re-added is fixed
- collection condition matches items on all levels: top-level items, attachments/notes, and annotations. So one can, for example, get all PDF attachments or annotations in a collection with just two conditions.
- "Include parent and child items of matching items" should include all ancestors or descendents of a search match. For example, if an annotation matches the query, "Include parent and child items of matching items" would make the attachment and the top-level item match too. Conversely, if a top-level item matches, with "Include parent and child items of matching items", attachment and annotations would match as well.
- finally, I ended up revising the optimization to avoid rerunning the entire search on `item` `modify` notifications. That way, one should be able to run a long-running query in a large library and then, say, tag an item without having to wait for another query to execute just to update a couple of rows.

One thing to note: while reworking the "Include parent and child items of matching items" search bit, I noticed that in `xpcom/data/search.js`, `includeParentsAndChildren` is found in two main places - `_buildQuery()` and `search()`. I updated `_buildQuery` to get this option to work with annotations in advanced search but I never found if/when this path in `search` would be relevant. I left it unchanged for now - what are the situation when this would be used, and should I update it similarly to `_buildQuery` to account for annotations too?

https://github.com/zotero/zotero/blob/d7d1d03b8befac715155d12200cca7b4cf50c498/chrome/content/zotero/xpcom/data/search.js#L770-L793

abaevbog commented:
https://forums.zotero.org/discussion/124008/advanced-search-what-is-the-meaning-of-does-not-contain it might be worth making this a separate issue. If one creates a condition "Annotation Comment does not contains X", we should probably only keep items from which annotation comment field is relevant.

But at least with this change, one will be able to also add "Item Type is Annotation" condition, which will yield only annotations without the "X" string.

AbeJellinek commented:
c1ad201c47f85092d0399e8c769257959c916dc7 merged as d2f1c56250cfbaeaef61dc3726d0ccfcaa0265ea
